src/script/descriptor.cpp

```diff
@@ -1087,17 +1087,36 @@ class WSHDescriptor final : public DescriptorImpl
     }
 };
 
+/** Represents the type of a node in taproot descriptor script tree */
+enum TRNodeType {
+    LEAF_SCRIPT,
+    NODE_HASH
+};
+
+/** A struct hold information on a node taproot descriptor script tree */
+struct TRNodeInfo {
+    int depth;
+    std::optional<int> leaf_version;
+    TRNodeType type;
+};
+
 /** A parsed tr(...) descriptor. */
 class TRDescriptor final : public DescriptorImpl
 {
-    std::vector<int> m_depths;
+    std::vector<TRNodeInfo> m_nodes;
 protected:
     std::vector<CScript> MakeScripts(const std::vector<CPubKey>& keys, Span<const CScript> scripts, FlatSigningProvider& out) const override
     {
         TaprootBuilder builder;
-        assert(m_depths.size() == scripts.size());
-        for (size_t pos = 0; pos < m_depths.size(); ++pos) {
-            builder.Add(m_depths[pos], scripts[pos], TAPROOT_LEAF_TAPSCRIPT);
+        assert(m_nodes.size() == scripts.size());
+        for (size_t pos = 0; pos < m_nodes.size(); ++pos) {
+            if (m_nodes[pos].type == TRNodeType::NODE_HASH) {
+                builder.AddOmitted(m_nodes[pos].depth, uint256(Span(scripts[pos])));
+            } else if (m_nodes[pos].type == TRNodeType::LEAF_SCRIPT) {
+                builder.Add(m_nodes[pos].depth, scripts[pos], m_nodes[pos].leaf_version.value());
```

Comment:
Having an `optional` when in one branch you ignore the value and in the other you assume it always has a value doesn't make a lot of sense to me.

Would it make sense to drop `leaf_version` from `TRNodeInfo` and store the leaf version as the first byte of the script? Alternatively, could just drop the optional and default it to 0.

Reply:
> Having an `optional` when in one branch you ignore the value and in the other you assume it always has a value doesn't make a lot of sense to me.
> 
> Would it make sense to drop `leaf_version` from `TRNodeInfo` and store the leaf version as the first byte of the script? Alternatively, could just drop the optional and default it to 0.

@ajtowns I prefer to drop the optional and default leaf_version to `0`